main/src/library/List.flix

```diff
@@ -0,0 +1,579 @@
+/*
+ * Copyright 2016 Liam Palmer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace List {
+
+    ///
+    /// Returns true iff `xs` is the empty list, i.e. `Nil`.
+    ///
+    def isEmpty[a](xs: List[a]): Bool = match xs with {
+        case Nil => true
+        case _ => false
+    }
+
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `xs`.
+    /// Returns `None` if `xs` is empty.
+    ///
+    def headOpt[a](xs: List[a]): Option[a] = match xs with {
+        case Nil => None
+        case x :: _ => Some(x)
+    }
+
+    ///
+    /// Returns `Some(x)` if `x` is the last element of `xs`.
+    /// Returns `None` if `xs` is empty.
+    ///
+    def lastOpt[a](xs: List[a]): Option[a] = match xs with {
+        case Nil => None
+        case x :: Nil => Some(x)
+        case x :: rs => lastOpt(rs)
+    }
+
+    ///
+    /// Returns the length of `xs`.
+    ///
+    def length[a](xs: List[a]): Int32 = match xs with {
+        case Nil => 0
+        case x :: rs => 1 + length(rs)
+    }
+
+    ///
+    /// Returns `ys` appended to `xs`.
+    ///
+    def append[a](xs: List[a], ys: List[a]): List[a] = match xs with {
+        case Nil => ys
+        case x :: rs => x :: append(rs, ys)
+    }
+
+    ///
+    /// Returns the element at index `i` in `xs`.
+    /// If `i < 0` or `i > length(xs)-1` the semantics are undefined.
+    ///
+    def at[a](i: Int32, xs: List[a]): a = match (i, xs) with {
+        case (0, x :: _) => x
+        case (p, x :: rs) => at(p - 1, rs)
+    }
+
+    ///
+    /// Returns `xs` with the first occurrence of `x` removed.
+    /// Returns `xs` if `x` does not exist in `xs`.
+    ///
+    def remove[a](x: a, xs: List[a]): List[a] = match xs with {
```

Comment:
If you could add the `@internal` annotation on all helper functions, then I will hide them in the flixdoc.